### PR TITLE
feat: Glitch-soc/Firefish emoji reactions API

### PR DESCRIPTION
**Pull request by ThatOneCalculator**
Adapted from https://git.joinfirefish.org/firefish/firefish/-/merge_requests/10532/diffs#diff-content-acabf6b2898e3102f138b148d469c390595935d1 from @VyrCossont

### CODE REVIEW

File: megalodon/src/entities/notification.ts [hunk 1]
```diff
@@ -7,7 +7,7 @@ namespace Entity {
     created_at: string
     id: string
     status?: Status
-    emoji?: string
+    reaction?: Reaction
```
Comment (h3poteto) [inline megalodon/src/entities/notification.ts:10]: Why did you change this parameter?

File: megalodon/src/entities/status.ts [hunk 2]
```diff
@@ -37,7 +37,7 @@ namespace Entity {
     application: Application | null
     language: string | null
     pinned: boolean | null
-    emoji_reactions: Array<Reaction>
+    reactions: Array<Reaction>
```
Comment (h3poteto) [inline megalodon/src/entities/status.ts:40]: Why did you change the name?

Comment (ThatOneCalculator) [reply megalodon/src/entities/status.ts:40]: More descriptive, given reactions can be unicode emojis, custom emojis, or technically even arbitrary text or images

File: megalodon/src/entities/notification.ts (hunk 1 above)
Comment (ThatOneCalculator) [reply megalodon/src/entities/notification.ts:10]: As per https://git.joinfirefish.org/firefish/firefish/-/blob/develop/packages/megalodon/src/entities/notification.ts?ref_type=heads#L10 , the reaction from a notification should be a `Reaction` object representing the actual reaction rather than just a string

Comment (h3poteto) [reply megalodon/src/entities/notification.ts:10]: But Pleroma provides only string as emoji in notifications: https://github.com/h3poteto/megalodon/blob/950ff623a3a0bdac46f07787a5256a6d200b391d/megalodon/src/pleroma/entities/notification.ts#L10
How to convert it?

Comment (ThatOneCalculator) [reply megalodon/src/entities/notification.ts:10]: Ahh... I'll have to look into that, but Glitch, Firefish, and Akkoma provide full strings. Furthermore, Pleroma is adding custom emoji reactions in its next update, so I think they'll end up changing that.

Comment (h3poteto) [reply megalodon/src/entities/notification.ts:10]: OK, please change this field after Pleroma adds custom emoji reactions.

File: megalodon/src/firefish.ts [hunk 3]
```diff
@@ -1303,6 +1303,59 @@ export default class Firefish implements MegalodonInterface {
...
+		if (isUnicodeEmoji) {
+			return name;
+		}
+		return `:${name}:`;
+	}
+
+	/**
+	 * POST /api/notes/reactions/create
+	 */
+	public async reactStatus(id: string, name: string): Promise<Response<Entity.Status>> {
```
Comment (h3poteto) [inline megalodon/src/firefish.ts:1322]: Please use `createEmojiReaction` method: https://github.com/h3poteto/megalodon/blob/950ff623a3a0bdac46f07787a5256a6d200b391d/megalodon/src/megalodon.ts#L1346

File: megalodon/src/entities/notification.ts (hunk 1 above)
Comment (ThatOneCalculator) [reply megalodon/src/entities/notification.ts:10]: How should we handle the difference between pleroma-style and glitch-style reactions for now?

File: megalodon/src/firefish.ts (hunk 3 above)
Comment (ThatOneCalculator) [reply megalodon/src/firefish.ts]: Also did for deleteEmojiReaction

File: megalodon/src/entities/status.ts (hunk 2 above)
Comment (ThatOneCalculator) [reply megalodon/src/entities/status.ts:40]: If you want, I can undo that to keep it as `emoji_reactions`

Comment (h3poteto) [reply megalodon/src/entities/status.ts:40]: No problem. However, since it is a breaking change, I'm wondering when to merge it.

File: megalodon/src/entities/notification.ts (hunk 1 above)
Comment (h3poteto) [reply megalodon/src/entities/notification.ts:10]: OK, please keep `emoji` field even though you add `reaction` field. After Pleroma supports custom emoji reactions, let's integrate it into `reaction` field.

File: megalodon/src/entities/status.ts (hunk 2 above)
Comment (ThatOneCalculator) [reply megalodon/src/entities/status.ts:40]: Hopefully sooner than later, I'd love to use upstream Megalodon inside Firefish :)

**Comment by h3poteto** [comment]
Please pass tests.

File: megalodon/src/entities/status.ts (hunk 2 above)
Comment (h3poteto) [reply megalodon/src/entities/status.ts:40]: Can you please move this change to another PR?
The field renaming should not be the main target of this PR. And if you separate it, I'll include it in the Firefish support release.

**Comment by ThatOneCalculator** [comment]
All checks passed!

File: megalodon/src/notification.ts [hunk 4]
```diff
@@ -5,6 +5,7 @@ namespace NotificationType {
   export const Favourite: Entity.NotificationType = 'favourite'
   export const Reblog: Entity.NotificationType = 'reblog'
   export const Mention: Entity.NotificationType = 'mention'
+  export const Reaction: Entity.NotificationType = 'reaction'
```
Comment (h3poteto) [inline megalodon/src/notification.ts:8]: Why did you add a new notification type? Is it used?

File: megalodon/src/firefish/api_client.ts [hunk 5]
```diff
@@ -266,7 +266,10 @@ namespace FirefishAPI {
           : '',
         plain_content: n.text ? n.text : null,
         created_at: n.createdAt,
-        emojis: Array.isArray(n.emojis) ? n.emojis.map(e => emoji(e)) : [],
+        // Remove reaction emojis with names containing @ from the emojis list.
+				emojis: Array.isArray(n.emojis) ? n.emojis
```
Comment (h3poteto) [inline megalodon/src/firefish/api_client.ts:270]: Please align the indent level.

File: megalodon/src/firefish/api_client.ts [hunk 6]
```diff
@@ -284,25 +287,34 @@ namespace FirefishAPI {
         application: null,
         language: null,
         pinned: null,
-        emoji_reactions: typeof n.reactions === 'object' ? mapReactions(n.reactions, n.myReaction) : [],
+        // Use emojis list to provide URLs for emoji reactions.
+				emoji_reactions: mapReactions(n.emojis!, n.reactions, n.myReaction),
```
Comment (h3poteto) [inline megalodon/src/firefish/api_client.ts:291]: Same as above.

File: megalodon/src/firefish/api_client.ts [hunk 7]
```diff
@@ -284,25 +287,34 @@ namespace FirefishAPI {
...
         bookmarked: false,
         quote: n.renote !== undefined && n.text !== null
       }
     }
 
-    export const mapReactions = (r: { [key: string]: number }, myReaction?: string | null): Array<MegalodonEntity.Reaction> => {
+    export const mapReactions = (emojis: Array<FirefishEntity.Emoji>, r: { [key: string]: number }, myReaction?: string | null): Array<MegalodonEntity.Reaction> => {
+      // Map of emoji shortcodes to image URLs.
+			const emojiUrls = new Map<string, string>(
```
Comment (h3poteto) [inline megalodon/src/firefish/api_client.ts:299]: Same as above.

File: megalodon/src/firefish/api_client.ts [hunk 8]
```diff
@@ -284,25 +287,34 @@ namespace FirefishAPI {
...
       return Object.keys(r).map(key => {
-        if (myReaction && key === myReaction) {
-          return {
-            count: r[key],
-            me: true,
-            name: key
-          }
-        }
+        // Strip colons from custom emoji reaction names to match emoji shortcodes.
+				const shortcode = key.replace(/:/g, "");
```
Comment (h3poteto) [inline megalodon/src/firefish/api_client.ts:304]: Same as above.

File: megalodon/src/firefish/api_client.ts [hunk 9]
```diff
@@ -284,25 +287,34 @@ namespace FirefishAPI {
...
+				// If this is a custom emoji (vs. a Unicode emoji), find its image URL.
+				const url = emojiUrls.get(shortcode);
+				// Finally, remove trailing @. from local custom emoji reaction names.
+				const name = shortcode.replace("@.", "");
```
Comment (h3poteto) [inline megalodon/src/firefish/api_client.ts:308]: Just confirm.
This logic means you want to remove `:` and @.` from emoji_reactions, right?
For example,
`{ count: 1, me: false, name: ':ablobfoxexplode@.:' }` -> `{ count: 1, me: false, name: 'ablobfoxexplode' }`

Comment (ThatOneCalculator) [reply megalodon/src/firefish/api_client.ts:308]: Yes

Comment (h3poteto) [reply megalodon/src/firefish/api_client.ts:308]: Thanks, I got it.

**Comment by ThatOneCalculator** [comment]
I'll be able to address the review items later today :)

File: megalodon/src/notification.ts (hunk 4 above)
Comment (ThatOneCalculator) [reply megalodon/src/notification.ts:8]: Yes, here: https://github.com/h3poteto/megalodon/pull/1921/files#diff-727a65323e887a8055dfc36e06244e7622dc446801b15f1027995e8b4c7cf63dR367

**Comment by ThatOneCalculator** [comment]
Requesting review from @h3poteto , test only fails due to the Friendica server chosen for the `detector` test being down (see https://github.com/h3poteto/megalodon/pull/1923/commits/afd1be25ad38a2887cd1f6bb7e768b7a6c3a28bf)

File: megalodon/src/notification.ts (hunk 4 above)
Comment (h3poteto) [reply megalodon/src/notification.ts:8]: Why did you add Reaction even though EmojiReaction exists? What's the difference?

Comment (ThatOneCalculator) [reply megalodon/src/notification.ts:8]: Mostly to stay in line with Firefish's `reaction` notification name. I don't think it's very necessary but helps with differentiating string-only reaction notifications (Pleroma) from Reaction object notifications

Comment (h3poteto) [reply megalodon/src/notification.ts:8]: Ah, I got it.
You mean
In Pleroma(current):
When we receive an EmojiReaction notification, it contains `emoji` field in the object (it does not contain `reaction`). And it should be `NotificationType.EmojiReaction`.
In Firefish:
When we receive a Reaction notification, it contains `reaction` field in the object (it does not contain `emoji`). And it should be `NotificationType.Reaction`.

Furthermore, in the future Pleroma, when we receive an EmojiReaction, it contains `reaction` field in the object instead of `emoji` field. And at that time, we can remove `NotificationType.EmojiReaction`.

Correct?

Comment (ThatOneCalculator) [reply megalodon/src/notification.ts:8]: Yep, you got it.

**Review by h3poteto — APPROVED** [review]
Thank you
